network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -182,13 +183,223 @@ func (t *GossipSubMeshTracer) Prune(p peer.ID, topic string) {
 	lg.Debug().Hex("flow_id", logging.ID(id.NodeID)).Str("role", id.Role.String()).Msg("pruned peer")
 }
 
-// SendRPC is called when a RPC is sent. Currently, the GossipSubMeshTracer tracks iHave RPC messages that have been sent.
+// SendRPC is called by GossipSub when a RPC is sent. Currently, the GossipSubMeshTracer tracks iHave RPC messages that have been sent.
 // This function can be updated to track other control messages in the future as required.
-func (t *GossipSubMeshTracer) SendRPC(rpc *pubsub.RPC, _ peer.ID) {
+func (t *GossipSubMeshTracer) SendRPC(rpc *pubsub.RPC, p peer.ID) {
 	err := t.rpcSentTracker.Track(rpc)
 	if err != nil {
 		t.logger.Err(err).Bool(logging.KeyNetworkingSecurity, true).Msg("failed to track sent pubsbub rpc")
 	}
+
+	msgCount, ihaveCount, iwantCount, graftCount, pruneCount := 0, 0, 0, 0, 0
+	if rpc.Control != nil {
+		ihaveCount = len(rpc.Control.Ihave)
+		iwantCount = len(rpc.Control.Iwant)
+		graftCount = len(rpc.Control.Graft)
+		pruneCount = len(rpc.Control.Prune)
+	}
+	msgCount = len(rpc.Publish)
+	t.metrics.OnRpcReceived(msgCount, ihaveCount, iwantCount, graftCount, pruneCount)
+	t.logger.Trace().
+		Str("remote_peer_id", p2plogging.PeerId(p)).
```

Comment:
A general comment about the methods in this file:

can/should this log message be gated on being in trace mode? `p2plogging.PeerId` seems to have some logic including serialization and looking up the id in a cache. Since these are very common code paths, I worry this will chew up CPU cycles unnecessarily. a simple approach is to just front load all of the metrics collection, and return early if `GetLevel()` returns trace.

Reply:
https://github.com/onflow/flow-go/pull/5083/commits/72d92bed34ad043b721dc2dd8993cad6cd1b19a6